### PR TITLE
Feat add merge of multiple opossums

### LINKED ISSUE #209: Accept multiple input files and merge them
We already set up the CLI to accept multiple input files. However until now, we just error if more than 1 file is provided.
Instead we should merge multiple input files into a single output file. For now we keep this simple and assume that all input files have the same root directory.

### PR DESCRIPTION
**Pull request by abraemer**
## Summary of changes
* added merge functionality
* changed CLI to accept multiple input files by merging them

## Context and reason for change

See #209 
Closes #209

### CODE REVIEW

File: src/opossum_lib/core/services/merge_opossums.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+        base_urls_for_sources=_merge_base_urls_for_sources(scan_results),
+        attribution_to_id=_merge_attribution_to_id(scan_results),
+        unassigned_attributions=unassigned_attributions_filtered,
+    )
+
+
+def _merge_metadata(scan_results: list[ScanResults]) -> Metadata:
+    return Metadata(
+        project_id=str(uuid.uuid4()),
+        project_title="Opossum Merger",
```
Comment (mstykow) [inline src/opossum_lib/core/services/merge_opossums.py:70]: i suppose this title is here because the PR is still draft. in fact, we'll have to do some validation. it doesn't really make sense to merge opossum files of different projects. we can talk.

File: src/opossum_lib/core/services/generate_impl.py [hunk 2]
```diff
@@ -4,13 +4,13 @@
...
+from opossum_lib.core.services.merge_opossums import merge_opossums
 from opossum_lib.core.services.write_opossum_file import write_opossum_file
 
 
 def generate_impl(input_readers: list[InputReader], output_file: Path) -> None:
-    # currently this converts only one file (validated in the arguments)
-    # for the future a merge step is planned after reading the files
-    opossum = input_readers[0].read()
+    opossums = [reader.read() for reader in input_readers]
+    opossum = opossums[0] if len(opossums) == 1 else merge_opossums(opossums)
```
Comment (mstykow) [inline src/opossum_lib/core/services/generate_impl.py:13]: it seems like this logic would better be done inside `merge_opossums`. as a user of this abstraction, i would expect that `merge_opossums` can handle all of its edge cases (such as nothing-to-merge because there is only 1 element) by itself. after all, i don't know the implementation of `merge_opossums` and neither should i need to. i would also expect that `merge_opossums` can handle the edge case of 0 elements by itself, or does it?

**Review by mstykow — COMMENTED** [review]
general comment: i see you introducing non-standard abbreviations throughout your code (val for value, attr for attribution, sr for scan result, ...). in my opinion this is an unfortunate habit because
a) every developer starts calling variables by their favorite abbreviation which leaves me guessing what each person means
b) it makes the code harder to read for no obvious gain. sure the line length might be shorter in some cases but i'd rather have something more explicit without needing to guess what you mean.

**Comment by abraemer** [comment]
While I agree that in general using non-standard abbreviations can make code harder to understand, I don't think this is necessarily the case here. Generally, I think the degree of detail a variable name should have can be correlated to the scope it is relevant for. The most extreme example are Python's list/dict comprehensions, where you introduce a variable that is only used in a single statement. This is where of the "nonstandard" variables occur in this PR.

Personally, I'd be totally fine with:
`return [attribution for attribution in unassigned_attributions if attribution not in all_attributions]`.
However this line is too long for the formatter (by like 10 characters) causing it to force linebreaks which makes the statement a lot less readable for me:
```
return [
    attribution
    for attribution in unassigned_attributions
    if attribution not in all_attributions
]
```
This just doesn't parse visually for me without additional mental effort. IMO, comprehensions should only be used as a short-hand for clear/easy operations and there they actually improve readability a lot. As soon as they would span multiple lines, I rewrite them as an explicit loop which is more readable to me than the line-wrapped comprehension:
```
unassigned_attributions = []
for attribution in attributions:
    if attribution not in all_attributions:
        unassigned_attributions.append(attribution)
return unassigned_attributions
```
But expresses the intend much more weakly then the original comprehension.

So most of these naming choices where driven by the formatter actually. I suppose this is not the best reason conceptually. Maybe we could increase the maximal allowed line width? Default is 88. I actually find myself fighting the formatter frequently because the line length always forces a trade_off between long and descriptive names and simple formatting.
If increasing the line length is not an option, then I'll consider living with the multi-line comprehensions or convert them to explicit loops.

File: tests/core/services/test_merge_opossums.py
Comment (Hellgartner) [inline tests/core/services/test_merge_opossums.py:1]: I would have liked some test of the actual behavior not only of the edge case

File: tests/core/services/test_merge_opossums.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+            review_results=empty_review_results.model_copy()
+        )
+        opossum2 = opossum_faker.opossum(
+            review_results=empty_review_results.model_copy()
+        )
+        opossum3 = opossum_faker.opossum(
+            review_results=empty_review_results.model_copy()
+        )
+        merged = merge_opossums([opossum1, opossum2, opossum3])
+        # Compare all field except metadata because there the UUID changes
```
Comment (Hellgartner) [inline tests/core/services/test_merge_opossums.py:34]: this comment seems wrong/misplaced

File: tests/input_formats/opossum/entities/generators/generate_outfile_information.py [hunk 4]
```diff
@@ -35,23 +45,103 @@ def output_file(
...
+        *,
+        resources: ResourceInFileModel | None = None,
+        external_attributions: dict[OpossumPackageIdentifierModel, OpossumPackageModel]
+        | None = None,
+        min_count: int = 1,
+        max_count: int = 5,
+        num_attributions: int = 3,
+    ) -> dict[str, list[str]]:
+        if resources is None:
+            resources = self.file_information_provider.resource_in_file()
```
Comment (Hellgartner) [inline tests/input_formats/opossum/entities/generators/generate_outfile_information.py:108]: IMHO this does not make a lot of sense without setting it in the resultfile as well

File: tests/input_formats/opossum/entities/generators/generate_outfile_information.py [hunk 5]
```diff
@@ -35,23 +45,103 @@ def output_file(
...
             or self.date_time_provider.date_time().isoformat(),
-            input_file_md5_checksum=input_file_md5_checksum
-            or entry_or_none(self.misc_provider, self.misc_provider.md5()),
+            input_file_md5_checksum=None,
         )
+
+    def manual_attributions(
+        self,
+    ) -> dict[str, ManualAttributions]:
+        return {}
```
Comment (Hellgartner) [inline tests/input_formats/opossum/entities/generators/generate_outfile_information.py:79]: is that desired that this function always returns a constant?

File: tests/input_formats/opossum/entities/generators/generate_outfile_information.py [hunk 6]
```diff
@@ -35,23 +45,103 @@ def output_file(
...
+                return paths
+            for path_segment, children in resources.items():
+                if isinstance(children, int):
+                    paths.append(path_segment)
+                else:
+                    subpaths = resources_to_path(children)
+                    paths.extend(path_segment + "/" + subpath for subpath in subpaths)
+            return paths
+
+        allpaths = ["/" + path for path in resources_to_path(resources)]
```
Comment (Hellgartner) [inline tests/input_formats/opossum/entities/generators/generate_outfile_information.py:131]: `allpaths` -> `all_paths`

File: tests/input_formats/opossum/entities/generators/generate_outfile_information.py [hunk 7]
```diff
@@ -35,23 +45,103 @@ def output_file(
...
+        how_many_paths = min(
```
Comment (Hellgartner) [inline tests/input_formats/opossum/entities/generators/generate_outfile_information.py:132]: `how_many` --> `number_of` (two times)

**Review by Hellgartner — CHANGES_REQUESTED** [review]

**Comment by mstykow** [comment]
> While I agree that in general using non-standard abbreviations can make code harder to understand, I don't think this is necessarily the case here. Generally, I think the degree of detail a variable name should have can be correlated to the scope it is relevant for. The most extreme example are Python's list/dict comprehensions, where you introduce a variable that is only used in a single statement. This is where of the "nonstandard" variables occur in this PR.
> 
> Personally, I'd be totally fine with: `return [attribution for attribution in unassigned_attributions if attribution not in all_attributions]`. However this line is too long for the formatter (by like 10 characters) causing it to force linebreaks which makes the statement a lot less readable for me:
> 
> ```
> return [
>     attribution
>     for attribution in unassigned_attributions
>     if attribution not in all_attributions
> ]
> ```
> 
> This just doesn't parse visually for me without additional mental effort. IMO, comprehensions should only be used as a short-hand for clear/easy operations and there they actually improve readability a lot. As soon as they would span multiple lines, I rewrite them as an explicit loop which is more readable to me than the line-wrapped comprehension:
> 
> ```
> unassigned_attributions = []
> for attribution in attributions:
>     if attribution not in all_attributions:
>         unassigned_attributions.append(attribution)
> return unassigned_attributions
> ```
> 
> But expresses the intend much more weakly then the original comprehension.
> 
> So most of these naming choices where driven by the formatter actually. I suppose this is not the best reason conceptually. Maybe we could increase the maximal allowed line width? Default is 88. I actually find myself fighting the formatter frequently because the line length always forces a trade_off between long and descriptive names and simple formatting. If increasing the line length is not an option, then I'll consider living with the multi-line comprehensions or convert them to explicit loops.

I don't think we want to avoid comprehensions just because they span multiple lines. I personally don't mind long comprehensions and just wouldn't even get into this conflict. I also don't mind making the line length longer, like 120 chars. Yes, there is a style guide out there with 80 chars but I don't see why we'd need to adhere to it if it causes us problems.

File: src/opossum_lib/core/services/generate_impl.py (hunk 2 above)
Comment (mstykow) [reply src/opossum_lib/core/services/generate_impl.py:13]: Probably best if we discuss this in a quick call.

File: src/opossum_lib/core/services/merge_opossums.py [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+
+def _has_nonempty_review_result(opossum: Opossum) -> bool:
+    return bool(
+        opossum.review_results
+        and (
+            opossum.review_results.manual_attributions
+            or opossum.review_results.resolved_external_attributions
+            or opossum.review_results.resources_to_attributions
+        )
+    )
```
Comment (mstykow) [inline src/opossum_lib/core/services/merge_opossums.py:81]: there are multiple issues here:
1. it's not common to use functional programming in python and in this case there is also no need to.
2. by using `filter` in combination with a filter function you're masking the real type of `opossum.review_results` which means you need to use a cast in line 70. but this, too, is not necessary.

the code becomes simpler when you filter the way you usually would inside a comprehension:
```suggestion
def _extract_review_results(opossums: list[Opossum]) -> list[OpossumOutputFileModel]:
    return [
        opossum.review_results
        for opossum in opossums
        if opossum.review_results
        and (
            opossum.review_results.manual_attributions
            or opossum.review_results.resolved_external_attributions
            or opossum.review_results.resources_to_attributions
        )
    ]
```

File: src/opossum_lib/core/services/merge_opossums.py [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+
+
+def _merge_scan_results(opossums: list[Opossum]) -> ScanResults:
+    scan_results = [opossum.scan_results for opossum in opossums]
+    resources = _merge_resources(scan_results)
+    unassigned_attributions = _merge_unassigned_attributions(scan_results)
+    unassigned_attributions_filtered = _filter_assigned_attributions(
```
Comment (mstykow) [inline src/opossum_lib/core/services/merge_opossums.py:88]: why is the variable called "unassigned attributions" when the function resulting in this variable is called "assigned attributions"?

**Review by mstykow — CHANGES_REQUESTED** [review]

File: src/opossum_lib/core/services/merge_opossums.py (hunk 9 above)
Comment (abraemer) [reply src/opossum_lib/core/services/merge_opossums.py:88]: "filter" in the sense of "filter out". So taking all attribution and removing the ones that are assigned, the unassigned ones remain. The function name probably could be clearer.

**Review by mstykow — APPROVED** [review]

**Review by Hellgartner — APPROVED** [review]
